src/dp_nat.c

```diff
@@ -823,6 +824,8 @@ void dp_del_all_neigh_nat_entries_in_vni(uint32_t vni)
 	for (item = TAILQ_FIRST(&nat_headp); item != NULL; item = tmp_item) {
 		tmp_item = TAILQ_NEXT(item, entries);
 		if ((item->vni == vni) || (vni == DP_NETWORK_NAT_ALL_VNI)) {
+			// can fail on non-presence (fine) or internal error (already logged)
```

Comment:
I was actually thinking what if the neigh nat linked list is empty and we still have an entry in dnat table ?
but this can not happen because the way `dp_del_vip_from_dnat` is implemented ? or would it be safer to call smt like this from `dp_free_vni_data` and do not touch `dp_dell_all_neigh_nat_entries_in_vni`

Disclaimer: Code below never run or tested.

```
int dp_del_dnat_by_vni(uint32_t vni)
{
    const void *key;
    void *data;
    uint32_t next = 0;
    int ret;
    struct nat_key *nkey;

    while ((ret = rte_hash_iterate(ipv4_dnat_tbl, &key, &data, &next)) >= 0) {
		if (DP_FAILED(ret))
			return DP_GRPC_ERR_ITERATOR;

        nkey = (struct nat_key *)key;
        if (nkey->vni == vni) {
            rte_free(data);

            if (DP_FAILED(rte_hash_del_key(ipv4_dnat_tbl, key))) {
                DPS_LOG_WARNING("Failed to delete DNAT key for IP: %u, VNI: %u\n", nkey->ip, nkey->vni);
            }
        }
    }

        return DP_GRPC_OK;
}

```

Reply:
The last time we discussed and looked at the details of the hashtable implementation. our conclusion was that it is safe to delete during iteration ? At least conntrack hashtable still does it that way. (and some other hashtables as well)